### DISCUSSION TITLE
Over UART update    (as a wired counterpart of Over the Air Update)

### DISCUSSION THREAD

**Discussion by jean-claudeF** [micropython/micropython#13530] [Ideas]
Hi,
I am wondering if I could update part of my program (mainly the settings) in the way that OTA (Over the air update) works.
(Why this? I find WiFi unreliable, but my Pico is outside the house, so it is not so easy to connect to the computer to reprogram.

I'm thinking about this:
1. In the main loop my program checks if there is something coming via RxD  (maybe some special code to init the update procedure)
2. When this happens the main program stops. Everythin on RxD is put into a string until a special end code is received.
3. The contents of the string is written to "settings.py"
4. The Pico does a reset.
5. The original program runs, only that the new settings are imported via "import settings"

What do you think of this?
Or have I overseen a simpler method?

**Comment by Josverl** [comment]
Im assuming that you only want to update config files stored on the VFS, not wanting to update the entire Firmware ?
Also what device would you use to connect from ? Another MCU, or a RPi perhaps ?

for just a config file,  you can enable the repl on a UART,  
With a rpi you could run mpremote to connect to the repl over UART to copy an updated config 

MCU to MCU transfers are also possible , but a bit more care and work is needed to create the "updater" 
to avoid missing a few bit / bytes and creating an incorrect config file.

**Reply by jean-claudeF** [reply]
Thank you very much!
I didn't know that REPL over UART was possible (why not, I don't know...).
If that's possible I know what to do, as I have already experience with communiction over USB:
[https://github.com/jean-claudeF/Picoconnect](url)

The only thing is:  I must stop the running program and switch to REPL.
Do I find that somewhere in the docs?

**Reply by jean-claudeF** [reply]
A quick search seems to reveal that I have to recompile Micropython to enable REPL on UART for the Pico:
https://stackoverflow.com/questions/77466427/pi-pico-w-micropython-how-to-enable-ctrlc-via-uart
while for the 8266 it is available as standard
?
Here is an interesting article that shows that it is possible on the PicoW:
https://joe.blog.freemansoft.com/2023/01/replicating-micropython-repl-across-hc.html
but not on the normal Pico

**Reply by jean-claudeF** [reply]
With the PicoW I succeeded now in getting a double REPL on UART and my PicoConnect on USB.
![serial_REPL](https://github.com/micropython/micropython/assets/26027162/84bce959-c460-4e27-8ef7-e208552839b9)
This is what I wanted, but only on the PicoW.

Something strange happens when I use Thonny instead of PicoConnect. There is a lot of rubbish coming together with the REPL text.

**Reply by Josverl** [reply]
Can you get a reliable repl on the uart with mpremote?

Thonny is nice  because it does a lot of stuff silently,   
and sometimes it's annoying, because it does a lot of stuff silently but different from what you expect

**Reply by jean-claudeF** [reply]
I have tested mpremote:
`mpremote connect /dev/ttyUSB0 `
and got the output of a running program (simply a counting loop generating output).
That was fine, BUT:
I could not stop the running program to do anything else but listening.
Have I overseen something?

**Reply by bixb922** [reply]
It's either REPL or a program running, except if you use aiorepl. That module allows a REPL while a program is running. For this to work, the running program must yield control using asyncio.
If you are not familiar with asyncio on Micropython, see this: https://github.com/peterhinch/micropython-async/blob/master/v3/README.md
(edit) I hope you can make that run over the UART. It certainly runs with the standard UART-over-USB

**Reply by jean-claudeF** [reply]
<It's either REPL or a program running >
I am concious of this fact. But I want to interrupt the running program   (Ctrl-C) to get the REPL.
This is easy in Thonny or Picoconnect via ACM0, but it does not work over USB0.
I rather would like to work without async, if possible.

**Reply by Josverl** [reply]
Well , [this issue](https://github.com/micropython/micropython/issues/1568)  was closed with [commit](https://github.com/micropython/micropython/commit/a60efa8202d84a2d3efe522009d700dac289538e) and  the remark that Ctrl-C should work on dupterm in UART , but that was on the stm32 port.

Seems logical that would be available on rp2 as well, but your results show that it is not

**Reply by andrewleech** [reply]
The "repl interrupt" handling is a special chunk of code that's constantly watching for the interrupt character (eg. Ctrl-c) on each interface. It hasn't always been added to _all_ interfaces though, sounds like it might be missing in rp2 uart. 

Instead of compiling with UART repl enabled, you might be able to use dupterm to add UART as an additional repl and then its interrupt handling might work?

Ah, I've just got those two links above working and see dupterm was already suggested there, but it's disabled for some reason on regular Pico. Seeing as you're already able to compile your own build you could turn on dupterm in your build too, though did you already test it on the Pico-w and find the interrupt doesn't work with dupterm either?

...  and I thought I'd seen something related to this recently; this feature was needed to enable repl interrupt in dupterm which was only merged for rp2 a couple of months ago: https://github.com/micropython/micropython/pull/13200
So ctrl-c should now work with uart-via-dupterm
I think as part of that same change dupterm should now be enabled by default on all rp2 boards: https://github.com/micropython/micropython/blob/master/ports%2Frp2%2Fmpconfigport.h#L103

**Reply by jean-claudeF** [reply]
Hmmm,
I'm not sure if I understand everything.
I had used "MicroPython v1.21.0 on 2023-10-06; Raspberry Pi Pico W with RP2040" on the PicoW where it worked without Ctrl-C.
Maybe it's stupid, but I didn't think of it being a version problem.
Do I have to compile it ( https://github.com/micropython/micropython/blob/master/ports%2Frp2%2Fmpconfigport.h#L103  ) myself or is there a readymade image?

**Reply by jean-claudeF** [reply]
I tried with my self compiled Micropython 1.23.0 with this result:
- output is echoed via dupterm
- no reaction to Ctrl - A/B/C/D

My solution is now this:

```python
import os
import time
from machine import UART, reset
uart = machine.UART(0, baudrate=115200)
os.dupterm(uart)             # This duplicates the REPL, but doesn't react to Ctrl-ABCD

i=0
while 1:
    
    # This checks for reset and stop:
    if uart.any():
        inp = uart.read()
        # stop running program:
        if b'\04' in inp:
            reset()
        if b'\03' in inp:
            break
      
    print(i)
    i+=1
    time.sleep(1)
```
